datapath: remove SNAT maps entries when kube-proxy is enabled

mazzy89 opened the pull request:
This path remove SNAT maps entries to support the case when the user toggles off from using BPF to kube-proxy

Fixes: #13835
Signed-off-by: Salvatore Mazzarino <salvatore@accuknox.com>

Please ensure your pull request adheres to the following guidelines:

- [x] For first time contributors, read [Submitting a pull request](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#submitting-a-pull-request)
- [x] All code is covered by unit and/or runtime tests where feasible.
- [x] All commits contain a well written commit description including a title,
      description and a `Fixes: #XXX` line if the commit addresses a particular
      GitHub issue.
- [x] All commits are signed off. See the section [Developer’s Certificate of Origin](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#dev-coo)
- [x] Provide a title or release-note blurb suitable for the release notes.
- [x] Thanks for contributing!

<!-- Description of change -->

Fixes: #issue-number

```release-note
Remove SNAT maps entries to support the case when the user toggles off from using BPF to kube-proxy.
```

brb commented on pkg/datapath/maps/map.go:
```diff
@@ -175,6 +175,10 @@ func (ms *MapSweeper) RemoveDisabledMaps() {
 		}...)
 	}
 
+	if option.Config.KubeProxyReplacement == option.KubeProxyReplacementDisabled {
```
This is not enough:
- The SNAT v4 map can be used by `EnableBPFMasquerade && EnableIPv4Masquerade`.
- It is not used when `option.Config.KubeProxyReplacement == option.KubeProxyReplacementPartial && option.Config.EnableNodePort == false`.

mazzy89 replied:
I've addressed the required changes.

brb reviewed (changes requested):
Thanks, some comments.

pchaigno commented on pkg/datapath/maps/map.go:
```diff
@@ -175,6 +175,22 @@ func (ms *MapSweeper) RemoveDisabledMaps() {
+
+	if !option.Config.EnableIPv4Masquerade {
+		maps = append(maps, []string{"cilium_snat_v4_external"}...)
+	}
```
This will attempt to remove the map twice if both `EnableBPFMasquerade` and `EnableIPv4Masquerade` are false. I think it can be merge into a single if condition. The same is true with the if statements below. Could you merge them?

mazzy89 replied:
> This will attempt to remove the map twice if both EnableBPFMasquerade and EnableIPv4Masquerade are false

that's true but I thought this won't be a problem because the function does not error out if the path of the map is not found in the filesystem.

mazzy89 replied:
I wired the requested changes. It should be more concise. thanks @pchaigno

pchaigno commented on pkg/datapath/maps/map.go:
```diff
@@ -175,6 +175,15 @@ func (ms *MapSweeper) RemoveDisabledMaps() {
+	}
+
+	if (option.Config.KubeProxyReplacement == option.KubeProxyReplacementDisabled) ||
+		(option.Config.KubeProxyReplacement == option.KubeProxyReplacementPartial && !option.Config.EnableNodePort) {
```
@brb Would it be enough to check `!option.Config.EnableNodePort` here given it is disabled if `option.Config.KubeProxyReplacement == option.KubeProxyReplacementDisabled`?

brb replied:
Yep, that's a good suggestion. Also, BPF masquerade depends on it, so it's enough to check only for NodePort.

mazzy89 replied:
Going to change that. Thanks guys.

Commit:
datapath: remove SNAT maps entries when kube-proxy is enabled

This path remove SNAT maps entries to support the case when the user toggles off from using BPF to kube-proxy.

Fixes: #13835
Signed-off-by: Salvatore Mazzarino <salvatore@accuknox.com>

brb reviewed (approved):
Thanks!
